src/python/pants/backend/python/lint/bandit/rules_integration_test.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import List, Optional
+
+from pants.backend.python.lint.bandit.rules import BanditTarget
+from pants.backend.python.lint.bandit.rules import rules as bandit_rules
+from pants.backend.python.lint.bandit.subsystem import Bandit
+from pants.backend.python.rules import download_pex_bin, pex
+from pants.backend.python.rules.pex import CreatePex
+from pants.backend.python.subsystems import python_native_code, subprocess_environment
+from pants.backend.python.subsystems.python_native_code import PythonNativeCode
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEnvironment
+from pants.backend.python.targets.python_library import PythonLibrary
+from pants.build_graph.address import Address
+from pants.build_graph.build_file_aliases import BuildFileAliases
+from pants.engine.fs import FileContent, InputFilesContent, Snapshot
+from pants.engine.legacy.structs import PythonTargetAdaptor
+from pants.engine.rules import RootRule
+from pants.engine.selectors import Params
+from pants.rules.core.lint import LintResult
+from pants.source.wrapped_globs import EagerFilesetWithSpec
+from pants.testutil.option.util import create_options_bootstrapper
+from pants.testutil.subsystem.util import init_subsystems
+from pants.testutil.test_base import TestBase
+
+
+class BanditIntegrationTest(TestBase):
+  good_source = FileContent(path="test/good.py", content=b"hashlib.sha256()\n")
+  bad_source = FileContent(path="test/bad.py", content=b"hashlib.md5()\n")
+  # MD5 is a insecure hashing function
+
+  @classmethod
+  def alias_groups(cls) -> BuildFileAliases:
+    return BuildFileAliases(targets={'python_library': PythonLibrary})
+
+  @classmethod
+  def rules(cls):
+    return (
+      *super().rules(),
+      *bandit_rules(),
+      *download_pex_bin.rules(),
+      *pex.rules(),
+      *python_native_code.rules(),
+      *subprocess_environment.rules(),
+      RootRule(CreatePex),
+      RootRule(Bandit),
+      RootRule(BanditTarget),
+      RootRule(PythonSetup),
+      RootRule(PythonNativeCode),
+      RootRule(SubprocessEnvironment),
```

Comment:
You should be able to remove all of these except for `BanditTarget`.

Also, we didn't talk about `RootRule`s last week. A `RootRule` is something that can act at the "root" of the rule graph, rather than needing to be computed via some other rule. Generally, we use it to set up the "instrinsics" (aka builtins) of the V2 engine, such as some of our filesystem abstractions like `Digest`:

https://github.com/pantsbuild/pants/blob/73c6e7b253ed2a4fc73cafc92dda4e7778be1731/src/python/pants/engine/fs.py#L277-L286 

In tests, we add use `RootRule` because we're not setting up the full graph used in production. Normally, `pants/backend/python/lint/python_lint_target.py` would provide a rule to "reach" `BanditTarget`, but `python_lint_target.py` is not registered in the test's rule graph, so we must register `BanditTarget` as a root.

Here, we can remove most these `RootRules` because using `OptionsBootstrapper` will now provide a way to "reach" these subsystems like `PythonSetup`.

(I might not be using the most precise language. CC other reviewers if you can clarify anything).

Reply:
That makes a ton of sense, thank you for explaining the under-the-hood mechanics!